### PR TITLE
Stricter typescript config

### PR DESCRIPTION
**Pull request by ElectronicBlueberry**
Enable `noUncheckedIndexedAccess` to improve type safety for possibly undefined object values.

Example:

`noUncheckedIndexedAccess` disabled

```ts
type NumberDictionary = {
    [key: string]: number;
}

const object: NumberDictionary = {
    keyA: 1
};

// type of a is number, but it is clearly undefined here
const a = object.keyB;
```

`noUncheckedIndexedAccess` enabled

```ts
type NumberDictionary = {
    [key: string]: number;
}

const object: NumberDictionary = {
    keyA: 1
};

// type of a is number | undefined, improving type safety
const a = object.keyB;
```

Refactors some affected files to account for these changes.

## How will this affect the code base?

Since it's still up to discussion as to whether or not include this change, I wanted to add a write-up on how exactly this change will affect front-end code, and what reasons there are for / against this change.

The Pros and Cons of this change can be summarized as follows:

**Pro**
- Catches more potential problems at compile time
- Enforces detailed type definitions
- Ensures better error messages, by highlighting the parts of code where a potential error may occur, and forcing a branch or assertion

**Con**
- Additional learning curve for typescript

With this change enabled, you will notice quite a bit of new errors. Many of these are potential crashes, unspecific or even wrong types, or potential places for error messages. Here are a few examples I encountered, and how to deal with them:

---

### Incorrect Warning

Typescript isn't always right when it comes to these warnings. Sometimes the surrounding code makes it obvious that an undefined value is not possible here. For example:

(This code is modified for this example)

```ts
if (row.length > 0) {
    // Try to split by comma first
    let rowDataSplit = row[0].split(","); // possible undefined error
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
    // Try to split by tab
    rowDataSplit = row[0].split("\t"); // possible undefined error
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
}
```

Solution 1: Reordering code. To tell typescript that we are sure this value exists, we can simply reorder the code.
This is the preferable solution, as it leads to cleaner code.

```ts
const firstElement = row[0];

if (firstElement) {
    // Try to split by comma first
    let rowDataSplit = firstElement.split(",");
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
    // Try to split by tab
    rowDataSplit = firstElement.split("\t");
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
}
```

Solution 2: Tell typescript you know better. We can override typescripts assumption using `!`. **Warning:** Only use this if the immediately surrounding code makes it obvious this value exists!

```ts
if (row.length > 0) {
    // Try to split by comma first
    let rowDataSplit = row[0]!.split(",");
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
    // Try to split by tab
    rowDataSplit = row[0]!.split("\t");
    if (rowDataSplit.length === num_columns) {
        return rowDataSplit;
    }
}
```

Notice the `!`, behind the element which typescript thought may be undefined.

Sometimes the first solution is not possible, like when we are checking for a specific length. This was the case in the code this example was inspired by. In this case, we must use the second solution:

```ts
[...]
    } else if (row.length === 1) {
        // Try to split by comma first
        let rowDataSplit = row[0]!.split(",");
        if (rowDataSplit.length === num_columns) {
            return rowDataSplit;
        }
        // Try to split by tab
        rowDataSplit = row[0]!.split("\t");
        if (rowDataSplit.length === num_columns) {
            return rowDataSplit;
        }
[...]
```

---

### Unspecific Types

Sometimes this error can be resolved, by simply being stricter about your types. The two main use-cases for this are `tuples` and `readonly objects`. These can also be defined as `arrays` and `records`, however, unlike `tuples` and `readonly objects`, these types have no guarantee that a value exists behind a specific index, which is where the typescript error message comes from.

**Readonly Objects**

```ts
const operatorForAlias: Record<string, string> = {
    lt: "<",
    le: "<=",
    ge: ">=",
    gt: ">",
    eq: ":",
};

export function getOperatorForAlias(alias: string): string {
    return operatorForAlias[alias];
}
```

The keys and values of constant object never change. Typescript does not know this, and throws a possible undefined error in the function. This can be solved with more specific types:

```ts
const operatorForAlias = {
    lt: "<",
    le: "<=",
    ge: ">=",
    gt: ">",
    eq: ":",
} as const;

type OperatorForAlias = typeof operatorForAlias;
type Alias = keyof OperatorForAlias;
type Operator = OperatorForAlias[Alias];

export function getOperatorForAlias(alias: Alias): Operator {
    return operatorForAlias[alias];
}
```

`as const` tells typescript the values will never change. The other types restrict the possible inputs and outputs of the function to the objects keys and values respectively, giving us much more detailed type hints, and resolving the error.

To ensure future code changes keep the object as a flat key-value string pair, we can further add `satisfies Record<string, string>` to it, as follows:

```ts
const operatorForAlias = {
    lt: "<",
    le: "<=",
    ge: ">=",
    gt: ">",
    eq: ":",
} as const satisfies Record<string, string>;
```

**Tuples**

Consider an array of arrays as follows:

```ts
const validAliases = [
    [">", "_gt"],
    ["<", "_lt"],
];

for (const [alias, substitute] of validAliases) {
    // ...
}
```

The types of `alias` and `substitute` will be `string | undefined`. This is because typescript resolved the type of `validAliases` to `string[][]`: an array of string arrays. The length of the inner string array is unknown here. It could contain 0, 2, or 20 elements, as far as typescript knows, so this type is correct. We however know, that the inner array will always be exactly two strings long.

This is known as a tuple. `[">", "_gt"]` can be restricted to a tuple by setting it's type to `[string, string]`: An array with exactly two elements, the first of type string, the second one also of type string. Tuples do not need to be two elements long, and do not need the same type on every index. E.g a tuple of `[string, string, number]` is perfectly valid.

To make the types of `alias` and `substitute` resolve to `string`, we can change the code as follows:

```ts
const defaultValidAliases: Array<[string, string]> = [
    [">", "_gt"],
    ["<", "_lt"],
];
```

This also restricts what elements can be inserted into this array, further improving type safety.

---

### Potential Errors

Typescript also catches potential errors, which have not been thrown / asserted. These are valid potential error sources, which can not be worked around as in the examples above. Properly asserting them, gives us much more helpful error messages, to track down the source of an issue.

```ts
const outputStep = this.stepStore.getStep(connection.output.stepId);
let terminalSource = outputStep.outputs.find((output) => output.name === connection.output.name);
```

This code will compile with errors, because `outputStep` is potentially undefined. While we may know that this state is not possible, a future error or mistake may still make it possible and properly asserting can be helpful. Here I'm using the new `assertDefined` utility.

```ts
const outputStep = this.stepStore.getStep(connection.output.stepId);
assertDefined(outputStep, `No such step with id ${connection.output.stepId}`);

let terminalSource = outputStep.outputs.find((output) => output.name === connection.output.name);
```

The code will now compile fine.

---

### Defaulting and Optional Chaining

Setting sensible default values, or using optional chaining can also be a way of dealing with possible undefined values in some situations.

Example

```ts
// if otherVar is undefined or null, myVar will be set to 4
const myVar = otherVar ?? 4;

// someFunction will only be called if someObject is not undefined or null
someObject?.someFunction();

// combining both concepts
const myOtherVar = someObject?.getValue() ?? "default";
```

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

**Comment by bgruening** [comment]
That is a great writeup @ElectronicBlueberry - I wonder if it should also go to some developer docs.

File: client/src/components/Workflow/Editor/modules/terminals.test.ts [hunk 1]
```diff
@@ -302,35 +302,7 @@ describe("canAccept", () => {
...
-        );
-        // disconnect, map over should be maintained
-        multiSimpleInputOne.disconnect(collectionOut);
-        expect(multiSimpleInputOne.localMapOver.isCollection).toBe(true);
-        expect(multiSimpleInputOne.canAccept(simpleDataOut).canAccept).toBe(false);
-        expect(multiSimpleInputOne.canAccept(simpleDataOut).reason).toBe(
-            "Cannot attach non-collection output to mapped over input, consider disconnecting inputs and outputs to reset this input's mapping."
-        );
-    });
+    // TODO: test mapOver reset when constraint removed
```
Comment (mvdbeek) [inline client/src/components/Workflow/Editor/modules/terminals.test.ts:334]: I think something went wrong with the merge here ?

File: client/src/components/Workflow/Editor/modules/terminals.ts [hunk 2]
```diff
@@ -273,9 +278,14 @@ class BaseInputTerminal extends Terminal {
     _collectionAttached() {
         const outputTerminals = this._getOutputTerminals();
         return outputTerminals.some((outputTerminal) => {
-            const output = this.stepStore
-                .getStep(outputTerminal.stepId)
-                .outputs.find((output) => output.name == outputTerminal.name);
+            const step = this.stepStore.getStep(outputTerminal.stepId);
+
+            if (!step) {
+                console.error(`Invalid step. Could not fine step with id ${outputTerminal.stepId} in store.`);
```
Comment (mvdbeek) [inline client/src/components/Workflow/Editor/modules/terminals.ts]:
```suggestion
                console.error(`Invalid step. Could not find step with id ${outputTerminal.stepId} in store.`);
```

**Comment by ElectronicBlueberry** [comment]
Test failures are unrelated.

File: client/src/stores/workflowConnectionStore.ts [hunk 3]
```diff
@@ -49,6 +48,20 @@ interface TerminalToInputTerminals {
     [index: string]: InputTerminal[];
 }
 
+/**
+ * Pushes a value to an array in an object, if the array exists. Else creates a new array containing value.
+ * @param object Object which contains array
+ * @param key Key which array is in
+ * @param value Value to push
+ */
+function pushOrSet<T>(object: { [key: string | number]: Array<T> }, key: string | number, value: T) {
```
Comment (mvdbeek) [inline client/src/stores/workflowConnectionStore.ts:57]: This seems generally useful, should it go to utils/ ?

**Review by mvdbeek — APPROVED** [review]
This is great, let's get this in before it goes stale again!

**Pull request merged**